### ISSUE TITLE
Panic: assignment to entry in nil map

### ISSUE THREAD

**Issue by wildermesser** [zalando-incubator/kube-metrics-adapter#379]
## Expected Behavior
Kube-metrics-adapter collects pod's metrics that are exposed as json.

## Actual Behavior
Kube-metrics-adapter fails with panic:
```
I1221 05:01:06.779152       1 serving.go:341] Generated self-signed cert (apiserver.local.config/certificates/apiserver.crt, apiserver.local.config/certifi 
 cates/apiserver.key)                                                                                                                                        
 time="2021-12-21T05:01:24Z" level=info msg="Looking for HPAs" provider=hpa                                                                                  
 E1221 05:01:24.973771       1 reflector.go:138] pkg/mod/k8s.io/client-go@v0.21.5/tools/cache/reflector.go:167: Failed to watch *v1.ClusterScalingSchedule:  
 failed to list *v1.ClusterScalingSchedule: the server could not find the requested resource (get ClusterScalingSchedules.zalando.org)                       
 E1221 05:01:24.976839       1 reflector.go:138] pkg/mod/k8s.io/client-go@v0.21.5/tools/cache/reflector.go:167: Failed to watch *v1.ScalingSchedule: failed  
 to list *v1.ScalingSchedule: the server could not find the requested resource (get ScalingSchedules.zalando.org)                                            
 time="2021-12-21T05:01:26Z" level=info msg="Adding new metrics collector: *collector.PodCollector" provider=hpa                                             
 time="2021-12-21T05:01:26Z" level=info msg="Found 5 new/updated HPA(s)" provider=hpa                                                                        
 E1221 05:01:26.261965       1 reflector.go:138] pkg/mod/k8s.io/client-go@v0.21.5/tools/cache/reflector.go:167: Failed to watch *v1.ClusterScalingSchedule:  
 failed to list *v1.ClusterScalingSchedule: the server could not find the requested resource (get ClusterScalingSchedules.zalando.org)                       
 E1221 05:01:26.581969       1 reflector.go:138] pkg/mod/k8s.io/client-go@v0.21.5/tools/cache/reflector.go:167: Failed to watch *v1.ScalingSchedule: failed  
 to list *v1.ScalingSchedule: the server could not find the requested resource (get ScalingSchedules.zalando.org)                                            
 time="2021-12-21T05:01:27Z" level=info msg="Collected 2 new metric(s)" provider=hpa                                                                         
 time="2021-12-21T05:01:27Z" level=info msg="Collected new custom metric 'core-cpu-load-level' (0) for Pod ptaf-core/ptaf-core-rs-00000000-0000-0000-0000-00 
 0000000000-6779d9f5cn46h" provider=hpa                                                                                                                      
 time="2021-12-21T05:01:27Z" level=info msg="Collected new custom metric 'core-cpu-load-level' (0) for Pod ptaf-core/ptaf-core-rs-00000000-0000-0000-0000-00 
 0000000000-6779d9f5v294r" provider=hpa                                                                                                                      
 panic: assignment to entry in nil map                                                                                                                       
 goroutine 79 [running]:                                                                                                                                     
 github.com/zalando-incubator/kube-metrics-adapter/pkg/provider.(*MetricStore).insertCustomMetric(_, {{{0x0, 0x0}, {0x0, 0x0}}, {{0x1e58b6a, 0x3}, {0xc0002c 
 3860, 0x9}, {0xc0007f8b00, ...}, ...}, ...})                                                                                                                
     /workspace/pkg/provider/metric_store.go:190 +0x58b                                                                                                      
 github.com/zalando-incubator/kube-metrics-adapter/pkg/provider.(*MetricStore).Insert(_, {{0xc000598f34, 0x4}, {0xc000598d30, 0x9}, {{{0x0, 0x0}, {0x0, 0x0} 
 }, {{0x1e58b6a, ...}, ...}, ...}, ...})                                                                                                                     
     /workspace/pkg/provider/metric_store.go:70 +0x98                                                                                                        
 github.com/zalando-incubator/kube-metrics-adapter/pkg/provider.(*HPAProvider).collectMetrics(0xc0002b7a40, {0x2208a70, 0xc00040ae00})                       
     /workspace/pkg/provider/hpa.go:261 +0x798                                                                                                               
 created by github.com/zalando-incubator/kube-metrics-adapter/pkg/provider.(*HPAProvider).Run                                                                
     /workspace/pkg/provider/hpa.go:100 +0x156                                                                                                               
 stream closed
```

## Steps to Reproduce the Problem

  1. Pods expose metrics: 
  ```
  curl http://$POD_IP:54322/metrics/json
{
  "core": {
    "load_level": 0
  }
  ```
  2. HPA configuration:
  ```
apiVersion: autoscaling/v2beta2
kind: HorizontalPodAutoscaler
metadata:
  annotations:
    autoscaling.alpha.kubernetes.io/behavior: '{"ScaleUp":{"StabilizationWindowSeconds":120,"SelectPolicy":"Max","Policies":[{"Type":"Pods","Value":4,"PeriodSeconds":15},{"Type":"Percent","Value":100,"PeriodSeconds":15}]},"ScaleDown":{"StabilizationWindowSeconds":null,"SelectPolicy":"Max","Policies":[{"Type":"Percent","Value":100,"PeriodSeconds":15}]}}'
    metric-config.pods.core-cpu-load-level.json-path/json-key: "$.core.load_level"
    metric-config.pods.core-cpu-load-level.json-path/path: /metrics/json
    metric-config.pods.core-cpu-load-level.json-path/port: "54322"
    metric-config.pods.core-cpu-load-level.json-path/scheme: "http"
  labels:
    tenant_id: 00000000-0000-0000-0000-000000000000
  name: hpa-00000000-0000-0000-0000-000000000000
  namespace: ptaf-core
spec:
  maxReplicas: 30
  minReplicas: 1
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: ptaf-core-rs-00000000-0000-0000-0000-000000000000
  metrics:
  - type: Pods
    pods:
      metric:
        name: core-cpu-load-level
      target:
        averageValue: 80
        type: AverageValue
  ```
## Specifications

  - Version: docker image kube-metrics-adapter@sha256:5be8869d21a9fb16bfeb1462e014439b4be81706c8a370d48d3a69db34694243
  - Platform: k8s 1.20.11

**Comment by szuecs** [comment]
@mikkeloscar we should also do something about the unrelated error logs for obviously missing `v1.ClusterScalingSchedule`.

**Comment by mikkeloscar** [comment]
@wildermesser I can't figure out which version of the kube-metrics-adapter you use based on the image hash?

What if you use the latest image: `registry.opensource.zalan.do/teapot/kube-metrics-adapter:v0.1.17-3-g2ccd690` Do you still see this issue?

**Comment by wildermesser** [comment]
Yes, I use:
```
crictl pull registry.opensource.zalan.do/teapot/kube-metrics-adapter:v0.1.17-3-g2ccd690
Image is up to date for sha256:d4809cfec2323d8cb74997ac5f4984fb503d1b364949956364e25ad2d7ceb077
```
and see this issue.
Sorry, I specified the wrong sha256 digest in the topic.

**Comment by mikkeloscar** [comment]
@wildermesser Thanks for the update. I tried to replicate the issue but didn't succeed so far. Do you have other HPAs in the cluster or do you only have one using `type: Pods` metrics?

**Comment by wildermesser** [comment]
There are some "standard" HPA's in the cluster with resource metrics like `autoscaling.alpha.kubernetes.io/metrics: '[{"type":"Resource","resource":{"name":"cpu","targetAverageValue":"800m"}}]'`. They are working correct because they collect metrics from "standart" metrics-server from k8s.gcr.io/metrics-server

**Comment by mikkeloscar** [comment]
@wildermesser Thanks, in the meantime we have also managed to replicate the problem and are looking more into the issue this week.

**Comment by wildermesser** [comment]
I confirm that issue is fixed in `kube-metrics-adapter:v0.1.17-6-g9d8359b`. Thank you, @katyanna
